io/src/main/scala/fs2/io/udp/AsynchronousSocketGroup.scala

```diff
@@ -32,55 +35,93 @@ object AsynchronousSocketGroup {
 
   def apply(): AsynchronousSocketGroup = new AsynchronousSocketGroup {
 
+    class Timeout(val expiry: Long, onTimeout: () => Unit) {
+      private var done: Boolean = false
+      def cancel: Unit = done = true
+      def timedOut: Unit = {
+        if (!done) {
+          done = true
+          onTimeout()
+        }
+      }
+    }
+
+    object Timeout {
+      def apply(duration: FiniteDuration)(onTimeout: => Unit): Timeout = new Timeout(System.currentTimeMillis + duration.toMillis, () => onTimeout)
+      implicit val ordTimeout: Ordering[Timeout] = Ordering.by[Timeout, Long](_.expiry)
+    }
+
     class Attachment(
-      readers: MutableQueue[Either[Throwable,Packet] => Unit] = MutableQueue(),
-      writers: MutableQueue[(Packet,Option[Throwable] => Unit)] = MutableQueue()
+      readers: ArrayDeque[(Either[Throwable,Packet] => Unit,Option[Timeout])] = new ArrayDeque(),
+      writers: ArrayDeque[((Packet,Option[Throwable] => Unit),Option[Timeout])] = new ArrayDeque()
     ) {
 
-      def hasReaders: Boolean = readers.nonEmpty
+      def hasReaders: Boolean = !readers.isEmpty
 
       def peekReader: Option[Either[Throwable,Packet] => Unit] = {
         if (readers.isEmpty) None
-        else Some(readers.head)
+        else Some(readers.peek()._1)
       }
 
       def dequeueReader: Option[Either[Throwable,Packet] => Unit] = {
         if (readers.isEmpty) None
-        else Some(readers.dequeue())
+        else {
+          val (reader, timeout) = readers.pop()
+          timeout.foreach(_.cancel)
+          Some(reader)
+        }
       }
 
-      def queueReader(reader: Either[Throwable,Packet] => Unit): Unit = {
-        if (closed) reader(Left(new ClosedChannelException))
-        else {
-          readers += reader
-          ()
+      def queueReader(reader: Either[Throwable,Packet] => Unit, timeout: Option[Timeout]): () => Unit = {
+        if (closed) {
+          reader(Left(new ClosedChannelException))
+          timeout.foreach(_.cancel)
+          () => ()
+        } else {
+          val r = (reader, timeout)
+          readers.add(r)
+          () => { readers.remove(r); () }
         }
       }
 
-      def hasWriters: Boolean = writers.nonEmpty
+      def hasWriters: Boolean = !writers.isEmpty
```

Comment:
It used to be `nonEmpty` but I had to change to a Java `ArrayDeque` for decent asymptotic random removal (on timeout expiration). The Java collection doesn't have a `nonEmpty`.